core/src/main/scala/org/apache/spark/MapOutputTracker.scala

```diff
@@ -1022,101 +1022,105 @@ private[spark] class MapOutputTrackerMaster(
-                totalOutputSize += blockSize
-              }
-            }
-            mapIdx = mapIdx + 1
-          }
-          val topLocs = locs.filter { case (loc, size) =>
-            size.toDouble / totalOutputSize >= fractionThreshold
-          }
-          // Return if we have any locations which satisfy the required threshold
-          if (topLocs.nonEmpty) {
-            return Some(topLocs.keys.toArray)
-          }
-        }
-      }
-    }
-    None
+  def getPreferredLocationsForShuffle(
+      dep: ShuffleDependency[_, _, _],
+      startMapIndex: Int,
+      endMapIndex: Int,
+      startReducerIndex: Int,
+      endReducerIndex: Int): Seq[String] = {
+    getLocationsWithLargestOutputs(
+      dep,
+      startMapIndex,
+      endMapIndex,
+      startReducerIndex,
+      endReducerIndex,
+      REDUCER_PREF_LOCS_FRACTION).map(_.host)
   }
 
   /**
-   * Return the locations where the Mappers ran. The locations each includes both a host and an
-   * executor id on that host.
+   * If shuffle locality is enabled, return a list of locations that each have fraction of
+   * map output greater than the specified threshold.
    *
    * @param dep shuffle dependency object
-   * @param startMapIndex the start map index
+   * @param startMapIndex the start map index (inclusive)
    * @param endMapIndex the end map index (exclusive)
-   * @return a sequence of locations where task runs.
+   * @param startReducerIndex the start reducer index (inclusive)
+   * @param endReducerIndex the end reducer index (exclusive)
+   * @param fractionThreshold fraction of total map output size that a location must have
+   *                          for it to be considered large.
    */
-  def getMapLocation(
+  def getLocationsWithLargestOutputs(
       dep: ShuffleDependency[_, _, _],
       startMapIndex: Int,
-      endMapIndex: Int): Seq[String] =
-  {
-    val shuffleStatus = shuffleStatuses.get(dep.shuffleId).orNull
-    if (shuffleStatus != null) {
-      shuffleStatus.withMapStatuses { statuses =>
-        if (startMapIndex < endMapIndex &&
-          (startMapIndex >= 0 && endMapIndex <= statuses.length)) {
-          val statusesPicked = statuses.slice(startMapIndex, endMapIndex).filter(_ != null)
-          statusesPicked.map(_.location.host).toSeq
-        } else {
-          Nil
+      endMapIndex: Int,
+      startReducerIndex: Int,
+      endReducerIndex: Int,
+      fractionThreshold: Double): Seq[BlockManagerId] = {
+    if (shuffleLocalityEnabled && dep.rdd.getNumPartitions < SHUFFLE_PREF_MAP_THRESHOLD &&
+      dep.partitioner.numPartitions < SHUFFLE_PREF_REDUCE_THRESHOLD) {
+      val shuffleStatus = shuffleStatuses.get(dep.shuffleId).orNull
+      if (shuffleStatus != null) {
+        shuffleStatus.withMapStatuses { statuses =>
+          if (statuses.nonEmpty && startMapIndex < endMapIndex &&
+            startMapIndex >= 0 && endMapIndex <= statuses.length) {
```

Comment:
The additional checks on `startMapIndex` and `endMapIndex` can be done outside along with the outer `shuffleLocalityEnabled` if condition (leave only the `endMapIndex <= statuses.length` here).